dask_mongo/core.py

```diff
@@ -14,19 +18,57 @@
 
 appname = f"dask-mongo/{__version__}"
 
+_CACHE_SIZE = 16
+
+
+def _recursive_tupling(item):
+    if isinstance(item, list):
+        return tuple([_recursive_tupling(i) for i in item])
+    if isinstance(item, Mapping):
+        return tuple(
+            [(_recursive_tupling(k), _recursive_tupling(v)) for k, v in item.items()]
+        )
+    else:
+        return item
+
+
+class _FrozenKwargs(dict):
+    def __hash__(self):
+        return hash(
+            frozenset(
+                [
+                    (_recursive_tupling(k), _recursive_tupling(v))
+                    for k, v in self.items()
+                ]
+            )
+        )
+
+
+@lru_cache(_CACHE_SIZE, typed=True)
+def _cache_inner(kwargs):
+    return pymongo.MongoClient(appname=appname, **kwargs)
+
+
+def _get_client(kwargs):
+    global _CLIENTS
+    frozen_kwargs = _FrozenKwargs(kwargs)
+    client = _cache_inner(frozen_kwargs)
+    atexit.register(weakref.WeakMethod(client.close))
```

Comment:
Suggest moving `atexit.register` to _cache_inner to avoid registering close for the same client many times.